feature: 인증 및 JWT 관련 기능 구현 및 리팩토링

jhyngu opened the pull request:
## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
- 인증 및 JWT 관련 기능을 구현하고 Spring Security를 활용한 인증 및 예외 처리 구조를 개선하였습니다.

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->
Commit: feat: DB 기반 로그인 검증 로직 구현
- 사용자 인증을 위한 DB 연동 로그인 검증 로직 추가.

Commit: feat: JWTUtil 구현 (RB-101), feat: JWTFilter 구현 (RB-135)
- JWTUtil 구현: JWT 생성, 만료 확인, 인증 정보 추출 등의 기능 포함.
- JWTFilter 구현: HTTP 요청에서 JWT를 검증하여 인증 정보를 SecurityContext에 저장.

Commit: feat: Refresh Rotate / Reissue 구현 (RB-136)
- Refresh Token을 Redis에 저장하고, 재발급 시 기존 토큰 폐기.

Commit: config: SecurityConfig 구현 (RB-138), refactor: 세션 생성 비활성화 설정 추가
- Spring Security 설정 파일 구현 (SecurityConfig).
- 세션 생성 비활성화 및 커스텀 필터 추가 설정.

Commit: feat: 로그인 필터 구현 (RB-99), feat: 로그아웃 필터 구현 (RB-137)
- LoginFilter와 LogoutFilter 구현: 로그인 요청 처리 및 로그아웃 시 토큰 무효화.

Commit: feat: AuthException 추가, refactor: GlobalExceptionHandler에 AuthException 처리 로직 추가
- AuthException 추가: 인증 관련 커스텀 예외 처리.
- GlobalExceptionHandler에 인증 예외 처리 로직 통합.

Commit: refactor: RoleType Enum 개선 및 권한 문자열 관리 방식 변경, refactor: CustomUserDetails 권한 반환 로직 간소화, chore: RefreshEntity 삭제
- RoleType Enum 구조 개선 및 권한 관리 로직 간소화.
- Redis를 활용한 Refresh Token 저장 및 관리.
- 불필요한 RefreshEntity 삭제.

Commit: feat: LoginRequest에 유효성 검증 어노테이션 추가
- LoginRequest에 유효성 검증 어노테이션 추가.

## ✅ 피드백 반영사항  <!-- 지난 코드리뷰에서 고친 사항을 적어주세요. 재PR 시에만 사용해 주세요! (재PR 아닌 경우 삭제) -->

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
- JWTFilter의 SecurityContext 저장 방식 : SecurityContextHolder에 인증 정보를 저장하는 과정에서 추가적으로 개선할 부분이 있는지 검토  부탁드립니다.
- 현재까지 테스트된 주요 시나리오 외에 놓친 테스트 케이스가 있다면 조언 부탁드립니다.

jhyngu committed:
feat: DB 기반 로그인 검증 로직 구현
- CustomUserDetails 구현
- CustomUserDetailsService 구현
- UserRepository findByEmail() 메소드 추가
- User Entity @Setter 추가

jhyngu committed:
feat: JWTUtil 구현 (RB-101)

jhyngu committed:
feat: JWTFilter 구현 (RB-135)

jhyngu committed:
feat: Refresh Rotate / Reissue 구현 (RB-136)
- RefreshEntity 구현
- RefreshRepository 구현
- ReissueController 구현
- ReissueService 구현

jhyngu committed:
config: SecurityConfig 구현 (RB-138)

jhyngu committed:
feat: 로그인 필터 구현 (RB-99)
- LoginFilter 구현
- LoginRequest 구현

jhyngu committed:
feat: 로그아웃 필터 구현 (RB-137)
- CustomLogoutFilter 구현

jhyngu committed:
refactor: 세션 생성 비활성화 설정 추가
- application.yml에 세션 생성 비활성화 설정 추가
- 불필요한 JSESSIONID 생성을 방지

jhyngu committed:
refactor: RoleType Enum 개선 및 권한 문자열 관리 방식 변경

- RoleType에 fullRole 필드 추가 (권한 문자열 포함)
- CustomUserDetails에서 권한 생성 방식 개선
- Spring Security의 권한 접두사 "ROLE_" 처리 로직 단순화

jhyngu committed:
refactor: Auth 관련 API 경로 변경

- 로그인 경로를 /auth/login으로 변경
- 로그아웃 경로를 /auth/logout으로 변경
- 토큰 갱신 경로를 /auth/reissue로 변경
- 보안 및 API 명확성을 위한 경로 개선

jhyngu committed:
feat: 로그인필터 에러 처리

- LoginFilter의 에러 처리 로직을 ErrorResponseUtil로 분리
- BusinessException 및 인증 실패에 대한 JSON 응답 생성을 유틸리티 클래스에서 처리
- 필터 내 코드 간결화 및 재사용성 향상
- GlobalExceptionHandler와의 일관된 에러 처리 구조 확립

jhyngu committed:
feat: JWT 발급, 갱신, 로그아웃 예외 처리 추가

- 토큰 발급 실패에 대한 상세 예외 처리 추가
- `ErrorResponseUtil`을 활용한 표준화된 에러 응답 구현
- 쿠키와 토큰 검증 로직 강화
- 로그아웃 중 발생할 수 있는 에러 처리 로직 추가

jhyngu committed:
refactor: CustomUserDetails 권한 반환 로직 간소화 및 유지보수성 향상

- 권한 반환 메서드(getAuthorities)를 간결하게 수정 (List.of() 활용)

jhyngu committed:
feat: LoginRequest에 유효성 검증 어노테이션 추가

- @Email 어노테이션을 사용하여 이메일 형식 검증 추가
  - "이메일 형식이 맞지 않습니다."라는 사용자 메시지 제공
- @NotBlank 어노테이션으로 필수 입력 필드 검증 추가
  - 이메일 및 비밀번호가 비어있을 경우 사용자 메시지 제공

jhyngu committed:
feat: 토큰 저장 방식을 Redis로 변경

- 기존의 메모리 기반 토큰 저장 방식을 Redis로 대체.
- `RefreshTokenService`를 추가하여 Redis와 상호작용하도록 구현.
- TTL을 적용하여 리프레시 토큰 자동 만료 지원.
- 관련 필터와 서비스 로직 수정.

jhyngu committed:
feat: 인증 처리용 커스텀 예외 클래스 추가

- 만료된 Access Token 처리를 위한 ExpiredAccessTokenException 추가
- 만료된 Refresh Token 처리를 위한 ExpiredRefreshTokenException 추가
- 잘못된 Refresh Token 처리를 위한 InvalidRefreshTokenException 추가
- 잘못된 토큰을 처리하는 InvalidTokenException 추가
- Refresh Token 미존재 처리를 위한 RefreshTokenNotFoundException 추가
- 토큰 재발급 실패 처리를 위한 TokenReissueFailedException 추가

jhyngu committed:
feat: AuthException 추가

- 인증 및 토큰 관련 예외 처리를 위한 AuthException 클래스 추가
- ErrorCode를 기반으로 메시지와 상태 코드를 처리
- 기존 BusinessException에서 분리하여 인증 도메인에 특화된 예외 처리 구현

jhyngu committed:
refactor: BusinessException 클래스 롤백

- 인증 도메인 예외 처리 분리로 인해 BusinessException은 본래의 역할로 복원
- 공통 비즈니스 로직 예외 처리를 위한 기존 구조로 유지
- 인증 및 토큰 관련 예외는 AuthException으로 대체

jhyngu committed:
refactor: GlobalExceptionHandler에 AuthException 처리 로직 추가

- 인증 및 토큰 관련 예외에 대해 AuthException 전용 핸들러 적용
- BusinessException은 비즈니스 로직 관련 예외로 처리 범위 명확화
- 각 예외에 적합한 HTTP 상태 코드와 메시지 반환

jhyngu committed:
refactor: 인증 관련 클래스에 커스텀 예외 적용

- CustomLogoutFilter: AuthException 대신 커스텀 예외 클래스 적용
  (ExpiredRefreshTokenException, InvalidTokenException 등)
- JWTFilter: 토큰 관련 예외를 커스텀 예외로 변환해 처리
  (ExpiredAccessTokenException, InvalidTokenException 등)
- LoginFilter: 인증 실패 시 커스텀 예외로 예외 처리 통일
  (InvalidTokenException, TokenReissueFailedException 등)
- ReissueController: 토큰 갱신 실패 시 커스텀 예외 클래스 적용
  (RefreshTokenNotFoundException, TokenReissueFailedException 등)

- 커스텀 예외 적용으로 예외 처리 로직의 명확성과 확장성 개선

jhyngu committed:
chore: RefreshEntity 삭제

- Redis를 사용한 Refresh 토큰 관리로 전환됨에 따라 RefreshEntity 불필요
- 기존 JPA 엔티티를 더 이상 사용하지 않아 삭제 처리

leeys9423 commented on backend/src/main/java/com/redbox/global/util/error/ErrorResponseUtil.java:
```diff
@@ -0,0 +1,53 @@
+    public static void handleAuthenticationError(HttpServletResponse response, Exception exception) throws IOException {
+        ErrorResponse errorResponse;
+
+        if (exception.getCause() instanceof BusinessException) {
```
인증 관련 실패인데 BusinessException으로 비교하는게 맞나요????

jhyngu replied:
AuthException으로 수정하겠습니다!

leeys9423 commented on backend/src/main/java/com/redbox/domain/auth/controller/ReissueController.java:
```diff
@@ -0,0 +1,76 @@
+
+            return ResponseEntity.ok().build();
+
+        } catch (AuthException e) {
```
에러 핸들링은 서비스에서 하는게 좋아보입니다잇

jhyngu replied:
네 알겠습니다!

jhyngu committed:
refactor: 인증 실패 처리 로직에서 BusinessException을 AuthException으로 변경

- 인증 관련 예외 처리에 적합한 AuthException으로 수정

jhyngu committed:
refactor: ReissueController에서 예외 처리를 ReissueService로 이동

- ReissueService에서 비즈니스 로직과 관련된 예외 처리 로직을 통합
  - RefreshToken 관련 검증 로직과 예외 처리를 서비스 계층으로 이동
  - ReissueController는 요청 파싱, 응답 작성에만 집중하도록 단순화
- 커스텀 예외 활용
- 코드 재사용성을 높이고 컨트롤러-서비스 계층 간 역할 분리를 강화

leeys9423 commented on backend/src/main/java/com/redbox/domain/auth/service/RefreshTokenService.java:
```diff
@@ -0,0 +1,42 @@
+    // Refresh Token 저장
+    public void saveRefreshToken(String email, String refreshToken, long ttl) {
+        // Redis에 RefreshToken 저장 (key: RefreshToken, value: email)
+        redisTemplate.opsForValue().set(REFRESH_TOKEN_KEY, email, ttl, TimeUnit.MILLISECONDS);
```
redis의 키값이 refresh_token_key 라는 상수로 고정이 되는 문제가 있네요
redisTemplate.opsForValue().set(REFRESH_TOKEN_KEY, email, ttl, TimeUnit.MILLISECONDS); -> redisTemplate.opsForValue().set(refreshToken, email, ttl, TimeUnit.MILLISECONDS);
이렇게 바꾸고 상단에 REFRESH_TOKEN_KEY 필요없을거 같아요

leeys9423 replied:
이거 제가 수정해서 커밋찍어놨어요

jhyngu replied:
감사합니다!👍

leeys9423 committed:
feat: CORS 설정 추가
- configuration.setExposedHeaders(List.of("access")) 설정 추가를 하지 않으면 클라이언트에서 Response Body에 접근할 수 없음
- ReissueController 사용하지 않은 import 삭제

leeys9423 committed:
fix: 레디스에 같은 키 값으로 생성하는 오류 수정

leeys9423 commented on backend/src/main/java/com/redbox/domain/user/entity/User.java:
```diff
@@ -2,15 +2,13 @@
 import java.time.LocalDateTime;
 
 @Getter
+@Setter
```
Setter를 달지말고 사용하는 필드에만 set메서드를 구현해서 사용하시는게 좋아요

jhyngu replied:
이거 저번에 질문드렸던건데 반영을 못 한 것 같아요. 수정하겠습니다!

leeys9423 commented on backend/src/main/java/com/redbox/domain/auth/filter/JWTFilter.java:
```diff
@@ -0,0 +1,83 @@
+                    .email(email)
+                    .roleType(RoleType.valueOf(role)) // RoleType 설정
+                    .build();
+
```
실제 DB기반의 User 정보가 아닌 토큰기반의 정보로 User 객체를 생성하고 있어서 문제가 발생할수 있을거 같아요.
CustomUserDetails을 만들때 DB를 거쳐야 할 거 같아요.

제가 고쳐서 커밋 찍어 놓을게요. 나중에 한번 비교해보세요~

leeys9423 committed:
feat: SecurityContext에서 User 정보 조회 기능 개선
- CustomUserDetails에 User 엔티티와 userId getter 추가
- UserService에 현재 로그인한 사용자 조회 메서드 구현
- SecurityContextHolder를 통한 사용자 정보 접근 로직 중복 제거
- JWTFilter에서 실제 DB 사용자 정보를 조회하도록 수정

leeys9423 committed:
feat: SecurityContext에서 User 정보 조회 기능 개선
- CustomUserDetails에 User 엔티티와 userId getter 추가
- UserService에 현재 로그인한 사용자 조회 메서드 구현
- SecurityContextHolder를 통한 사용자 정보 접근 로직 중복 제거
- JWTFilter에서 실제 DB 사용자 정보를 조회하도록 수정

leeys9423 committed:
refactor: findByEmail Optional 처리

jhyngu committed:
refactor: User클래스 @Setter 제거

jhyngu committed:
Merge remote-tracking branch 'origin/feature/login' into feature/login

jhyngu committed:
Merge branch 'devBE' into feature/login
